host/cli/log.go

```diff
@@ -64,3 +87,87 @@ func getLog(hostID, jobID string, client *cluster.Client, follow, init bool, std
 	_, err = attachClient.Receive(stdout, stderr)
 	return err
 }
+
+type LogLine struct {
+	Token   int
+	Content []byte
+}
+
+type LogRing struct {
+	logLines []*LogLine
+	start    int
+}
+
+func NewLogRing(capacity int) *LogRing {
+	return &LogRing{
+		logLines: make([]*LogLine, 0, capacity),
+	}
+}
+
+func (r *LogRing) Add(l *LogLine) {
+	if len(r.logLines) < cap(r.logLines) {
+		r.logLines = append(r.logLines, l)
+	} else {
+		r.logLines[r.start] = l
+		r.start++
+
+		if r.start == cap(r.logLines) {
+			r.start = 0
+		}
+	}
+}
+
+func (r *LogRing) Read() []*LogLine {
+	buf := make([]*LogLine, len(r.logLines))
+	if n := copy(buf, r.logLines[r.start:len(r.logLines)]); n < len(r.logLines) {
+		copy(buf[n:], r.logLines[:r.start])
+	}
+	return buf
+}
+
+func scanLogs(reader io.Reader, token int, output chan LogLine) {
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		buf := make([]byte, len(scanner.Bytes())+len([]byte("\n")))
```

Comment:
The length of "\n" is always going to be 1 byte, and it's added two lines down, so I think `+1` should be clear enough.